### DISCUSSION TITLE
include only some languages on remark-highlight.js

### DISCUSSION THREAD

**Discussion by arturcarvalho** [remarkjs/remark#629] [Q&A] [answered]
Hi there,

I'm using remark-highlight.js and I'm trying to do a light build. The problem is the bundle analyzer is showing all the languages inside the bundle:

![image](https://user-images.githubusercontent.com/2487710/107991760-8ed66880-6fd7-11eb-996b-78022ba3aa8f.png)

This is my current configuration to process markdown:
```ts
import slug from 'remark-slug';
import headings from 'remark-autolink-headings';
import hint from 'remark-hint';
import gfm from 'remark-gfm';
import remark from 'remark';
import html from 'remark-html';
import hljs from 'remark-highlight.js';

export async function markdownToHtml(markdown: string) {
  const result = await remark()
    .use(gfm) 
    .use(slug) 
    .use(headings, { behavior: 'wrap' }) 
    .use(hint)
    .use(hljs, {include:["javascript"]}) // this doesn't work, the bundle stays the same and I lose the highlights
    .use(html)
    .process(markdown);
  return result.toString();
}

```

But I can't find a way to make it work. It always bundles everything and I can't find any "lost" import in the code.
Am I missing any configuration? Any help is appreciated.

Thanks and thanks for the great work with the remark plugins ecosystem!
Artur

**Comment by wooorm** [comment, marked as answer]
Correct, highlight.js and prism, and thus lowlight and refractor, which are used in these plugins, come bundled with all languages by default.

Perhaps [indutny/common-shake](https://github.com/indutny/common-shake) might help here.

Alternatively, lowlight and refractor both have ways to start without anything, and then manually register languages.

A PR that adds a `light` plugin to `rehype-highlight`, which uses `lowlight/lib/core`, and requires folks to pass in languages, would be a welcome addition

**Reply by arturcarvalho** [reply]
Thank you for the reply Titus! 

I'd like to help with the PR although I'm not sure I have enough knowledge:
- Can you point me to an example that has only a subset of languages working that I could use as basis to integrate with `reyhpe-highlight`?
- Are people using these plugins and loading 300KB of data even when they just need a fraction of it?
- I get a bit confused with remark/rehype/unified. Is there a preferred method to use?

Hope I'm not bothering too much. I'll research on my own but of course some pointers would help.

Have a nice day.

**Reply by wooorm** [reply]
No problem at all!

> 1. Can you point me to an example that has only a subset of languages working that I could use as basis to integrate with reyhpe-highlight?

https://github.com/wooorm/lowlight#browser. Here’s how I’d do it: you’d probably want a new entry in `rehype-highlight`, `light.js` or so. Then that, and `index.js`, will probably use a `core.js` file that does all the current work, but `index.js` will pass a full lowlight in, whereas `light.js` will pass an “empty” lowlight in.

> 2. Are people using these plugins and loading 300KB of data even when they just need a fraction of it?

Folks typically process their content on the server side I guess. There are also several highlighters out there, and several integration plugins, so it might be that some of them do support this.

> 3. I get a bit confused with remark/rehype/unified. Is there a preferred method to use?

remark is for markdown. rehype is for HTML. All are unified pipelines. Here we’re talking about adding `<span>`s with classes, so that’s HTML, not markdown. See also https://unifiedjs.com/learn/

**Reply by wooorm** [reply]
P.S. Historically there was only remark, that’s why there are copies of several plugins in both ecosystems. First I added notes at the top of readmes to point folks to the rehype versions. Soon I’ll go and deprecate the remark versions if the work is more suitable for rehype

**Reply by arturcarvalho** [reply]
Thanks for all the info! 

I'm going to go through it and see if I can grok it. Didn't know about the unifiedjs/learn link.
I suppose you meant "remark is for markdown".

I'm using this in a nextjs project and it only triggered the increase in the build because I started to use the highlight functionality in a place where the code is dynamic. But I'm using the same exact function to highlight code coming from the server and the highlight `span` is on the HTML. 
There's some step going on there that just gets rid of the languages because it doesn't need them anymore.

Anyway, great tips, thank you very much!
